clarity/src/vm/analysis/type_checker/contexts.rs

```diff
@@ -36,7 +37,108 @@ pub struct TypingContext<'a> {
     pub depth: u16,
 }
 
+enum TraitContext {
+    /// Traits stored in this context use the trait type-checking behavior defined in Clarity1
+    Clarity1(HashMap<ClarityName, BTreeMap<ClarityName, FunctionSignature>>),
+    /// Traits stored in this context use the new trait type-checking behavior defined in Clarity2
+    Clarity2 {
+        /// Aliases for locally defined traits and traits imported with `use-trait`
+        defined: HashSet<ClarityName>,
+        /// All traits which are defined or used in a contract
+        all: HashMap<TraitIdentifier, BTreeMap<ClarityName, FunctionSignature>>,
+    },
+}
+
+impl TraitContext {
+    pub fn new(clarity_version: ClarityVersion) -> TraitContext {
+        match clarity_version {
+            ClarityVersion::Clarity1 => Self::Clarity1(HashMap::new()),
+            ClarityVersion::Clarity2 => Self::Clarity2 {
+                defined: HashSet::new(),
+                all: HashMap::new(),
+            },
+        }
+    }
+
+    pub fn is_name_used(&self, name: &str) -> bool {
+        match self {
+            Self::Clarity1(map) => map.contains_key(name),
+            Self::Clarity2 { defined, all: _ } => defined.contains(name),
+        }
+    }
+
+    pub fn add_defined_trait(
+        &mut self,
+        contract_identifier: QualifiedContractIdentifier,
+        trait_name: ClarityName,
+        trait_signature: BTreeMap<ClarityName, FunctionSignature>,
+    ) -> CheckResult<()> {
+        match self {
+            Self::Clarity1(map) => {
+                map.insert(trait_name, trait_signature);
+            }
+            Self::Clarity2 { defined, all } => {
+                defined.insert(trait_name.clone());
+                all.insert(
+                    TraitIdentifier {
+                        name: trait_name,
+                        contract_identifier,
+                    },
+                    trait_signature,
+                );
+            }
+        }
+        Ok(())
+    }
+
+    pub fn add_used_trait(
+        &mut self,
+        alias: ClarityName,
+        trait_id: TraitIdentifier,
+        trait_signature: BTreeMap<ClarityName, FunctionSignature>,
+    ) -> CheckResult<()> {
+        match self {
+            Self::Clarity1(map) => {
+                map.insert(trait_id.name, trait_signature);
+            }
+            Self::Clarity2 { defined, all } => {
+                defined.insert(alias);
+                all.insert(trait_id, trait_signature);
+            }
+        }
+        Ok(())
+    }
+
+    pub fn get_trait(
+        &self,
+        trait_id: &TraitIdentifier,
+    ) -> Option<&BTreeMap<ClarityName, FunctionSignature>> {
+        match self {
+            Self::Clarity1(map) => map.get(&trait_id.name),
+            Self::Clarity2 { defined: _, all } => all.get(trait_id),
+        }
+    }
+
+    pub fn into_contract_analysis(&mut self, contract_analysis: &mut ContractAnalysis) {
```

Comment:
Miner style nit: `into_*` is usually used in Rust to indicate that the type is going to be _converted into_ a new type.  For example:

```
impl Foo {
   pub fn into_bar(self) -> Bar {
      /* ... */
   }
}
```

Can this method be renamed to something like `update_contract_analysis()` to indicate that this `TraitContext` is sticking around after the call, and that the given `contract_analysis` already exists and is only going to be mutated?  Thanks!

Reply:
The traits are being drained, and it's being called from `ContractContext::into_contract_analysis`, so that is why I went with this name. `update_` doesn't get the point across that it is draining the `TraitContext`. I'm drawing a blank on another name that would be better than these two...

Reply:
okay, the current name is fine with me.